File: backend/src/main/java/com/urbanman/ecommerce/controller/ReviewController.java
package com.urbanman.ecommerce.controller;

import com.urbanman.ecommerce.model.Product;
import com.urbanman.ecommerce.model.Review;
import com.urbanman.ecommerce.model.User;
import com.urbanman.ecommerce.repository.ReviewRepository;
import com.urbanman.ecommerce.service.ProductService;
import com.urbanman.ecommerce.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.*;
import org.springframework.lang.NonNull;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/reviews")
public class ReviewController {

    @Autowired private ReviewRepository reviewRepo;
    @Autowired private ProductService productService;
    @Autowired private UserService userService;

    // Get all reviews for a product with average rating
    @GetMapping("/product/{productId}")
    public ResponseEntity<?> getProductReviews(@PathVariable Long productId) {
        List<Review> reviews = reviewRepo.findByProductProductId(productId);
        Double avg = reviewRepo.avgRatingByProduct(productId);
        Map<String, Object> result = new HashMap<>();
        result.put("reviews", reviews);
        result.put("averageRating", avg != null ? Math.round(avg * 10.0) / 10.0 : 0);
        result.put("totalReviews", reviews.size());
        return ResponseEntity.ok(result);
    }

    // Submit a review
    @PostMapping
    public ResponseEntity<?> addReview(@RequestBody Map<String, Object> body) {
        long productId = Long.parseLong(body.get("productId").toString());
        long userId    = Long.parseLong(body.get("userId").toString());
        Integer rating = Integer.valueOf(body.get("rating").toString());
        String comment = body.getOrDefault("comment", "").toString();

        if (rating < 1 || rating > 5) {
            return ResponseEntity.badRequest().body("Rating must be between 1 and 5");
        }

        Product product = productService.getProductById(productId);
        User user       = userService.findById(userId);

        if (product == null || user == null) {
            return ResponseEntity.notFound().build();
        }

        // One review per user per product
        if (reviewRepo.existsByProductProductIdAndUserUserId(productId, userId)) {
            return ResponseEntity.status(HttpStatus.CONFLICT).body("You have already reviewed this product");
        }

        Review review = new Review();
        review.setProduct(product);
        review.setUser(user);
        review.setUserName(user.getName());
        review.setRating(rating);
        review.setComment(comment);

        return ResponseEntity.ok(reviewRepo.save(review));
    }

    // Delete a review
    @DeleteMapping("/{reviewId}")
    public ResponseEntity<String> deleteReview(@PathVariable @NonNull Long reviewId) {
        reviewRepo.deleteById(reviewId);
        return ResponseEntity.ok("Review deleted");
    }
}

File: backend/controllers/reviewController.js
const Review = require('../models/Review');
const Product = require('../models/Product');
const User = require('../models/User');

const getProductReviews = async (req, res) => {
  try {
    const { productId } = req.params;
    const reviews = await Review.findAll({
      where: { productId },
      order: [['createdAt', 'DESC']]
    });

    let sum = 0;
    reviews.forEach(r => {
      sum += r.rating;
    });

    const avg = reviews.length ? sum / reviews.length : 0;
    const averageRating = Math.round(avg * 10) / 10;

    return res.status(200).json({
      reviews,
      averageRating,
      totalReviews: reviews.length
    });
  } catch (error) {
    console.error('Get reviews error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

const addReview = async (req, res) => {
  try {
    const { productId, userId, rating, comment } = req.body;

    if (!productId || !userId || rating === undefined) {
      return res.status(400).json({ error: 'productId, userId, and rating are required' });
    }

    const ratingVal = parseInt(rating, 10);
    if (ratingVal < 1 || ratingVal > 5) {
      return res.status(400).json({ error: 'Rating must be between 1 and 5' });
    }

    const product = await Product.findByPk(productId);
    const user = await User.findByPk(userId);

    if (!product || !user) {
      return res.status(404).json({ error: 'Product or User not found' });
    }

    // One review per user per product
    const existingReview = await Review.findOne({ where: { productId, userId } });
    if (existingReview) {
      return res.status(409).json({ error: 'You have already reviewed this product' });
    }

    const review = await Review.create({
      productId,
      userId,
      userName: user.name,
      rating: ratingVal,
      comment: comment || ''
    });

    return res.status(200).json(review);
  } catch (error) {
    console.error('Add review error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

const deleteReview = async (req, res) => {
  try {
    const { reviewId } = req.params;
    const review = await Review.findByPk(reviewId);
    if (!review) {
      return res.status(404).json({ error: 'Review not found' });
    }
    await review.destroy();
    return res.status(200).send('Review deleted');
  } catch (error) {
    console.error('Delete review error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
};

module.exports = {
  getProductReviews,
  addReview,
  deleteReview
};
